crates/sui-storage/src/lock_service.rs

```diff
@@ -241,24 +251,34 @@ impl LockServiceImpl {
 
         for (i, lock) in locks.iter().enumerate() {
             // The object / version must exist, and therefore lock initialized.
-            let lock = lock.ok_or(SuiError::TransactionLockDoesNotExist)?;
-
-            if let Some(previous_tx_digest) = lock {
-                // Lock already set to different transaction
-                if previous_tx_digest != tx_digest {
+            let lock = lock.as_ref().ok_or(SuiError::TransactionLockDoesNotExist)?;
+
+            if let Some(LockInfo {
+                epoch: previous_epoch,
+                tx_digest: previous_tx_digest,
+            }) = lock
+            {
+                // Lock already set to different transaction from the same epoch.
+                // If the lock is set in a previous epoch, it's ok to override it.
+                if previous_epoch == &epoch && previous_tx_digest != &tx_digest {
                     // TODO: add metrics here
                     debug!(prev_tx_digest =? previous_tx_digest,
                           cur_tx_digest =? tx_digest,
-                          "Conflicting transaction!  Lock state changed in unexpected way");
+                          "Conflicting transaction! Lock state changed in unexpected way");
                     return Err(SuiError::ConflictingTransaction {
-                        pending_transaction: previous_tx_digest,
+                        pending_transaction: *previous_tx_digest,
                     });
                 }
-            } else {
-                // Only write the locks that need to be written (are uninitialized)
-                let obj_ref = owned_input_objects[i];
-                locks_to_write.push((obj_ref, Some(tx_digest)));
+                if &epoch == previous_epoch {
+                    // Exactly the same epoch and same transaction, nothing to lock here.
+                    continue;
+                } else {
+                    debug!(prev_epoch =? previous_epoch, cur_epoch =? epoch, ?tx_digest, "Overriding an old lock from previous epoch");
```

Comment:
Shouldn't we ensure that `epoch > prev_epoch` before overriding? What if an old transaction from a previous epoch which was still running for whatever reason try to overtake a lock from a tx in newer epoch?